File: src/main/java/io/github/gaming32/superpack/util/SimpleHttp.java
package io.github.gaming32.superpack.util;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URL;
import java.net.URLConnection;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.github.gaming32.superpack.Superpack;

public final class SimpleHttp {
    private static final Logger LOGGER = LoggerFactory.getLogger(SimpleHttp.class);

    public static final String USER_AGENT;

    static {
        final Package specPackage = Superpack.class.getPackage();
        String userAgent = specPackage.getSpecificationTitle() + '/' + specPackage.getSpecificationVersion();
        if (userAgent.equals("null/null")) {
            userAgent = Superpack.class.getSimpleName() + "/DEBUG";
        }
        USER_AGENT = userAgent;
    }

    private SimpleHttp() {
    }

    public static URL createUrl(String base, String route, Map<String, Object> queryParams) throws MalformedURLException {
        final StringBuilder url = new StringBuilder(base);
        if (!base.endsWith("/")) {
            url.append('/');
        }
        if (!route.startsWith("/")) {
            url.append(route);
        } else {
            url.append(route, 1, route.length());
        }
        return buildUrl(url, queryParams);
    }

    public static URL createUrl(String url, Map<String, Object> queryParams) throws MalformedURLException {
        return buildUrl(new StringBuilder(url), queryParams);
    }

    private static URL buildUrl(StringBuilder url, Map<String, ? extends Object> queryParams) throws MalformedURLException {
        if (!queryParams.isEmpty()) {
            url.append('?');
            for (final var entry : queryParams.entrySet()) {
                url.append(entry.getKey());
                url.append('=');
                url.append(URLEncoder.encode(entry.getValue().toString(), StandardCharsets.UTF_8));
                url.append('&');
            }
            url.setLength(url.length() - 1);
        }
        return new URL(url.toString());
    }

    public static URLConnection request(URL url) throws IOException {
        LOGGER.info("Requesting {}", url);
        final URLConnection cnxn = url.openConnection();
        cnxn.setRequestProperty("User-Agent", USER_AGENT);
        return cnxn;
    }

    public static InputStream stream(URL url) throws IOException {
        return request(url).getInputStream();
    }
}

File: src/main/kotlin/io/github/gaming32/superpack/util/SimpleHttp.kt
package io.github.gaming32.superpack.util

import org.slf4j.LoggerFactory
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL
import java.net.URLConnection
import java.net.URLEncoder

object SimpleHttp {
    private val LOGGER = LoggerFactory.getLogger(SimpleHttp::class.java)
    @JvmStatic
    val USER_AGENT: String = run {
        val specPackage = JavaUtil.SUPERPACK_CLASS.`package`
        val userAgent = "${specPackage.specificationTitle}/${specPackage.specificationVersion}"
        if (userAgent == "null/null") {
            JavaUtil.SUPERPACK_CLASS.simpleName + "/DEBUG"
        } else {
            userAgent
        }
    }

    @JvmStatic
    @Throws(MalformedURLException::class)
    fun createUrl(base: String, route: String, queryParams: Map<String, *>): URL {
        val url = StringBuilder(base)
        if (!base.endsWith("/")) {
            url.append('/')
        }
        if (!route.startsWith("/")) {
            url.append(route)
        } else {
            url.append(route, 1, route.length)
        }
        return buildUrl(url, queryParams)
    }

    @JvmStatic
    @Throws(MalformedURLException::class)
    fun createUrl(url: String, queryParams: Map<String, *>) = buildUrl(StringBuilder(url), queryParams)

    @JvmStatic
    @Throws(MalformedURLException::class)
    private fun buildUrl(url: StringBuilder, queryParams: Map<String, *>): URL {
        if (!queryParams.isEmpty()) {
            url.append('?')
            for ((key, value) in queryParams) {
                url.append(key)
                url.append('=')
                url.append(URLEncoder.encode(value.toString(), Charsets.UTF_8))
                url.append('&')
            }
            url.setLength(url.length - 1)
        }
        return URL(url.toString())
    }

    @JvmStatic
    @Throws(IOException::class)
    fun URL.request(): URLConnection {
        LOGGER.info("Requesting {}", this)
        val cnxn = openConnection()
        cnxn.setRequestProperty("User-Agent", USER_AGENT)
        return cnxn
    }

    @JvmStatic
    @Throws(IOException::class)
    fun URL.stream() = request().getInputStream()
}